### PR TITLE
MassDownloader: flexible string template

### PR DESCRIPTION
**Pull request by paride**
### What does this PR do?

Lets the mass downloader string template work without specifying all the fields.
Fixes: #2174. 

### PR Checklist
- [x] Correct base branch selected? `master` for new features, `maintenance_...` for bug fixes
- [ ] This PR is not directly related to an existing issue (which has no PR yet).
- [x] If the PR is making changes to documentation, docs pages can be built automatically.
      Just remove the space in the following string after the + sign: "+DOCS"
- [x] If any network modules should be tested for the PR, add them as a comma separated list
      (e.g. `clients.fdsn,clients.arclink`) after the colon in the following magic string: "+TESTS:clients.fdsn"
      (you can also add "ALL" to just simply run all tests across all modules)
- [ ] All tests still pass.
- [ ] Any new features or fixed regressions are be covered via new tests.
- [x] Any new or changed features have are fully documented.
- [ ] Significant changes have been added to `CHANGELOG.txt` .
- [x] First time contributors have added your name to `CONTRIBUTORS.txt`.

### CODE REVIEW

File: obspy/clients/fdsn/mass_downloader/utils.py [hunk 1]
```diff
@@ -542,9 +542,8 @@ def get_mseed_filename(str_or_fct, network, station, location, channel,
     if callable(str_or_fct):
         path = str_or_fct(network, station, location, channel, starttime,
                           endtime)
-    elif ("{network}" in str_or_fct) and ("{station}" in str_or_fct) and \
-            ("{location}" in str_or_fct) and ("{channel}" in str_or_fct) and \
-            ("{starttime}" in str_or_fct) and ("{endtime}" in str_or_fct):
+    elif any("{%s}" % key in str_or_fct for key in ("network", "station", \
+                "location", "channel", "starttime", "endtime")):
```
Comment (megies) [inline obspy/clients/fdsn/mass_downloader/utils.py]: please get rid of redundant backslash (inside brackets) and fix indent of continuation line, see circle CI with code check results

**Comment by paride** [comment]
@krischer thanks for your feedback. If you don't think this is a good idea we can rediscuss it. I did't think this as a problematic change: an ambiguity in the file name should be pretty clear to spot, but you certainly have a broader vision.

**Comment by krischer** [comment]
> @krischer thanks for your feedback. If you don't think this is a good idea we can rediscuss it. I did't think this as a problematic change: an ambiguity in the file name should be pretty clear to spot, but you certainly have a broader vision.

The problem I see is that the filename is used for a lot of things. The mass downloader is for example designed so that it can run multiple times and will only download missing data. If there is now ambiguity in the filenames it might wrongly assume that some data already exist and will not attempt to download it in the first place. This would be really hard to spot for users.

Now that I've thought a bit more about it: I think a better solution would be to (right at the start of the `.download()` method) check if the passed string is supposed to be a string template (e.g. check for the curly braces), and if yes: make sure all required fields and no extra fields are present. If that test fails, raise a nice and instructive errors message.

This is potentially a bit more hassle to use than your solution but I feel like it would avoid a lot of potential confusion.

**Comment by paride** [comment]
I'm closing this for the reasons given in @krischer's last comment.

**Pull request closed without merging**